### PR TITLE
AnyValueModel

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

File: Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    public:
+      using Type = ValueModel<AnyRef>::Type;
+
+      using UpdateSignal = ValueModel<AnyRef>::UpdateSignal;
+
+      /**
+       * Constructs an AnyValueModel over an existing ValueModel.
+       * @param source The ValueModel to wrap.
+       */
+      template<typename T>
```
Comment (spireken) [inline Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp:24]: Should these use Type instead of T?

**Review by spireken — COMMENTED** [review]
Looks good.

File: Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp (hunk 1 above)
Comment (shaobinspire) [reply Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp:24]: `Type` and `T` are different. `Type` of the AnyValueModel is `AnyRef`, while `T` used in the constructor could be of any type. Taking the unit test in AnyValueModelTester for example, T is `int`.

**Review by spireken — APPROVED** [review]

File: Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        const UpdateSignal::slot_type& slot) const override;
+
+    private:
+      mutable UpdateSignal m_update_signal;
+      Type m_value;
+      std::function<void (const Type&)> m_set_value;
+  };
+
+  template<typename T>
+  AnyValueModel::AnyValueModel(std::shared_ptr<ValueModel<T>> source)
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp:43]: ```source``` is a ```std::shared_ptr``` passed by value but no ownership of it is ever taken, that will result in a dangling reference.

File: Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    : m_value(const_cast<T&>(source->get())),
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp:44]: This violates const-correctness, can engender undefined behavior, and violates several of model's invariants, such as all mutations should result in a signal.

File: Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      m_set_value([=] (const Type& value) {
+        any_cast<T>(m_value) = any_cast<const T>(value); }) {}
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp:46]: This implementation leads to dangling references (```value```'s ownership is never expressed).

File: Applications/Spire/Source/Common/AnyValueModel.cpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+
+const AnyValueModel::Type& AnyValueModel::get() const {
+  return m_value;
+}
+
+QValidator::State AnyValueModel::test(const Type& value) const {
+  if(m_value.has_value()) {
```
Comment (kranar) [inline Applications/Spire/Source/Common/AnyValueModel.cpp:12]: ```value``` is never used, it's not clear what this method is doing.

**Review by kranar — CHANGES_REQUESTED** [review]
Too many issues with this, almost nothing about this implementation is correct. Revise it and resubmit a new PR.

**Pull request merged**